File: ios/robokassa_sdk/Sources/robokassa_sdk/RobokassaSdkPlugin.swift
import Flutter
import UIKit

/// Bridges Flutter to Robokassa's official iOS SDK.
///
/// That SDK is vendored into this same module under `RobokassaSDK/` rather than
/// linked as a separate pod, so there is no `import RobokassaSDK` — see
/// `RobokassaSDK/VENDORED.md`.
///
/// The SDK's `Robokassa` class presents its own `WebViewController` on top of
/// whatever is frontmost — under Flutter that is the `FlutterViewController` —
/// and reports through three closures. Those closures are *not* one-shot:
/// while the SDK polls the payment-state service it calls `onFailureHandler`
/// repeatedly for transient "not paid yet" answers, and it dismisses the screen
/// a couple of seconds after a terminal answer. Each flow is therefore wrapped
/// in a `PaymentSession` that settles its Pigeon callback exactly once and
/// remembers the last failure so a later dismissal can be classified.
public class RobokassaSdkPlugin: NSObject, FlutterPlugin, RobokassaHostApi {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = RobokassaSdkPlugin()
    RobokassaHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    registrar.publish(instance)
  }

  /// The in-flight checkout, retained so the SDK object outlives this call.
  private var session: PaymentSession?

  // ---------------------------------------------------------------------------
  // RobokassaHostApi
  // ---------------------------------------------------------------------------

  public func isNativeSdkAvailable() throws -> Bool {
    // The SDK is compiled into this module, so it cannot be absent.
    return true
  }

  public func startPayment(
    request: RkPaymentRequest,
    completion: @escaping (Result<RkPaymentResultMessage, Error>) -> Void
  ) {
    guard session == nil else {
      completion(.failure(PigeonError(
        code: "busy",
        message: "A Robokassa checkout is already in progress. Await the first "
          + "call before starting another.",
        details: nil)))
      return
    }

    if request.mode == .savedCard, (request.order.token ?? "").isEmpty {
      completion(.failure(PigeonError(
        code: "invalid_arguments",
        message: "A saved-card payment needs `order.token` — the opKey of the "
          + "earlier payment that saved the card.",
        details: nil)))
      return
    }

    let params: PaymentParams
    do {
      params = try RobokassaSdkPlugin.makePaymentParams(request)
    } catch {
      completion(.failure(error))
      return
    }

    let robokassa = Robokassa(
      login: request.credentials.merchantLogin,
      password: request.credentials.password1,
      password2: request.credentials.password2,
      isTesting: request.credentials.isTest)

    // Asked for on the success path so the result carries Robokassa's real
    // `<Result><Code>`/`<State><Code>`, as the Android SDK's success already
    // does. Skipped without an invoice: `checkPaymentParams` omits `invoiceID`
    // then, and the service has nothing to answer for.
    let checkBody = params.order.invoiceId > 0
      ? RobokassaSdkPlugin.makeCheckStatusBody(request, params)
      : nil

    let session = PaymentSession(
      robokassa: robokassa,
      invoiceId: request.order.invoiceId,
      checkBody: checkBody
    ) { [weak self] result in
      self?.session = nil
      completion(.success(result))
    }
    self.session = session
    session.attachHandlers()

    switch request.mode {
    case .simple:
      robokassa.startSimplePayment(with: params)
    case .hold:
      robokassa.startHoldingPayment(with: params)
    case .recurrent:
      robokassa.startDefaultReccurentPayment(with: params)
    case .savedCard:
      robokassa.startPaymentBySavedCard(with: params)
    }
  }

  public func confirmHold(
    request: RkPaymentRequest,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    run(request, completion) { robokassa, params, done in
      robokassa.confirmHoldingPayment(with: params, completion: done)
    }
  }

  public func cancelHold(
    request: RkPaymentRequest,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    run(request, completion) { robokassa, params, done in
      robokassa.cancelHoldingPayment(with: params, completion: done)
    }
  }

  public func chargeRecurring(
    request: RkPaymentRequest,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    guard let previous = request.order.previousInvoiceId, previous > 0 else {
      completion(.failure(PigeonError(
        code: "invalid_arguments",
        message: "chargeRecurring needs `order.previousInvoiceId` — the invoice "
          + "of the first, already-paid payment in the series.",
        details: nil)))
      return
    }
    run(request, completion) { robokassa, params, done in
      robokassa.startReccurentPayment(with: params, completion: done)
    }
  }

  public func checkPaymentState(
    request: RkPaymentRequest,
    completion: @escaping (Result<RkPaymentResultMessage, Error>) -> Void
  ) {
    // Deliberately *not* routed through the SDK's `ServiceCheckPaymentStatus`,
    // whose `checkPaymentStatus()` rebuilds its request body from a
    // `UserDefaults` key written by the last checkout — it would answer for
    // whichever invoice was paid most recently, not the one asked about, and a
    // wrong payment status is worse than no status. Its transport underneath,
    // `RequestManager.requestForCheckStatus`, takes the body as an argument and
    // has no such coupling, so this builds the body from *this* request and
    // calls that directly.
    let params: PaymentParams
    do {
      params = try RobokassaSdkPlugin.makePaymentParams(request)
    } catch {
      completion(.failure(error))
      return
    }

    // Both hoisted so the task below closes over nothing but `Sendable` values.
    let body = RobokassaSdkPlugin.makeCheckStatusBody(request, params)
    let invoiceId = request.order.invoiceId

    Task { @MainActor in
      do {
        let codes = try await RobokassaSdkPlugin.queryStateCodes(body: body)
        completion(.success(RobokassaSdkPlugin.makeStateResult(
          invoiceId: invoiceId,
          resultCode: codes.result,
          stateCode: codes.state)))
      } catch {
        completion(.failure(PigeonError(
          code: "robokassa_native_error",
          message: error.localizedDescription,
          details: nil)))
      }
    }
  }

  /// The `MerchantLogin:invoiceID:Password2` body for *this* request's invoice.
  ///
  /// `checkPaymentParams` signs with password #2, which the SDK's own entry
  /// points inject; reaching the transport directly means doing it here.
  fileprivate static func makeCheckStatusBody(
    _ request: RkPaymentRequest,
    _ params: PaymentParams
  ) -> String {
    var query = params
    query.merchantLogin = request.credentials.merchantLogin
    query.password2 = request.credentials.password2
    return query.checkPaymentParams
  }

  /// Robokassa's `<Result><Code>` and `<State><Code>` from one `OpStateExt` answer.
  fileprivate struct StateCodes {
    let result: String?
    let state: String?
  }

  /// Queries the payment-state service for `body`, bounded by a timeout.
  ///
  /// The bound matters: this also runs on the checkout's success path, where
  /// `URLSession`'s own 60s default would leave the Dart `await` pending long
  /// after the user is back in the app.
  fileprivate static func queryStateCodes(
    body: String,
    timeoutNanoseconds: UInt64 = 15_000_000_000
  ) async throws -> StateCodes {
    try await withThrowingTaskGroup(of: StateCodes?.self) { group in
      // `RequestManager.shared` is main-actor isolated.
      group.addTask { @MainActor in
        let xml = try await RequestManager.shared.requestForCheckStatus(to: body)
        return StateCodes(
          result: RequestManager.shared.getStateCode(from: xml, "Result"),
          state: RequestManager.shared.getStateCode(from: xml, "State"))
      }
      group.addTask {
        try await Task.sleep(nanoseconds: timeoutNanoseconds)
        return nil
      }
      // Whichever finishes first wins; the loser is cancelled and its result
      // discarded as the group drains on the way out.
      defer { group.cancelAll() }
      guard let codes = try await group.next() ?? nil else {
        throw MessagedError(
          message: "Timed out waiting for the Robokassa payment-state service.")
      }
      return codes
    }
  }

  /// Classifies an `OpStateExt` `<Result><Code>` / `<State><Code>` pair, matching
  /// how the Dart fallback maps the same endpoint.
  private static func makeStateResult(
    invoiceId: Int64?,
    resultCode: String?,
    stateCode: String?
  ) -> RkPaymentResultMessage {
    guard resultCode == "0" else {
      // Robokassa refused the query itself (bad signature, unknown invoice, …);
      // there is no payment state to report.
      let reason = PaymentResult(rawValue: resultCode ?? "") ?? .notFound
      return RkPaymentResultMessage(
        outcome: .error,
        invoiceId: invoiceId,
        resultCode: resultCode,
        stateDescription: reason.title,
        errorMessage: reason.title)
    }

    let state = PaymentState(rawValue: stateCode ?? "") ?? .notFound
    let outcome: RkPaymentOutcome
    switch state {
    case .paymentSuccess, .holdSuccess:
      outcome = .success
    case .cancelledNotPaid:
      outcome = .canceled
    default:
      outcome = .error
    }

    return RkPaymentResultMessage(
      outcome: outcome,
      invoiceId: invoiceId,
      resultCode: resultCode,
      stateCode: stateCode,
      stateDescription: state.title,
      errorMessage: outcome == .error ? state.title : nil)
  }

  // ---------------------------------------------------------------------------
  // Headless operations
  // ---------------------------------------------------------------------------

  /// Shared plumbing for the three completion-handler operations.
  private func run(
    _ request: RkPaymentRequest,
    _ completion: @escaping (Result<Bool, Error>) -> Void,
    _ body: (Robokassa, PaymentParams, @escaping (Result<Bool, Error>) -> Void) -> Void
  ) {
    let params: PaymentParams
    do {
      params = try RobokassaSdkPlugin.makePaymentParams(request)
    } catch {
      completion(.failure(error))
      return
    }

    let robokassa = Robokassa(
      login: request.credentials.merchantLogin,
      password: request.credentials.password1,
      password2: request.credentials.password2,
      isTesting: request.credentials.isTest)

    // Capture `robokassa` in the closure so it outlives this scope; the SDK
    // keeps no strong reference to itself while the request is in flight.
    body(robokassa, params) { result in
      withExtendedLifetime(robokassa) {
        switch result {
        case .success(let value):
          completion(.success(value))
        case .failure(let error):
          completion(.failure(PigeonError(
            code: "robokassa_native_error",
            message: error.localizedDescription,
            details: nil)))
        }
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Transport records -> Robokassa SDK models
  // ---------------------------------------------------------------------------

  private static func makePaymentParams(_ request: RkPaymentRequest) throws -> PaymentParams {
    // Hoisted so the `try` covers a whole statement rather than sitting inside
    // an argument list.
    let receipt = try makeReceipt(request.order.receipt)

    // Rejected rather than dropped: silently nilling an unrecognised currency
    // would charge the merchant's default one instead, which a caller could
    // only discover from their settlement report.
    var outSumCurrency: Currency?
    if let raw = request.order.outSumCurrency {
      guard let parsed = Currency(rawValue: raw.lowercased()) else {
        throw PigeonError(
          code: "invalid_arguments",
          message: "Unknown settlement currency \"\(raw)\".",
          details: nil)
      }
      outSumCurrency = parsed
    }

    var order = OrderParams(
      invoiceId: Int(request.order.invoiceId ?? -1),
      orderSum: request.order.orderSum,
      description: request.order.orderDescription,
      incCurrLabel: request.order.incCurrLabel,
      token: request.order.token,
      outSumCurrency: outSumCurrency,
      expirationDate: request.order.expirationDateEpochMs.map {
        Date(timeIntervalSince1970: TimeInterval($0) / 1000.0)
      },
      receipt: receipt)

    // Not covered by `OrderParams.init`, but public and settable.
    order.previousInvoiceId = Int(request.order.previousInvoiceId ?? -1)
    order.isRecurrent = request.order.isRecurrent
    order.isHold = request.order.isHold

    // The SDK spells English `eng`, so `en` has to be translated. Anything else
    // is rejected rather than coerced to English — a typo used to change the
    // checkout language with no diagnostic.
    var culture: Culture?
    if let raw = request.customer.culture {
      switch raw.lowercased() {
      case "ru": culture = .ru
      case "en", "eng": culture = .eng
      default:
        throw PigeonError(
          code: "invalid_arguments",
          message: "Unknown checkout language \"\(raw)\"; expected \"ru\" or \"en\".",
          details: nil)
      }
    }

    let customer = CustomerParams(
      culture: culture,
      email: request.customer.email,
      ip: request.customer.ip)

    let view = ViewParams(
      toolbarBgColor: request.view.toolbarBgColor,
      toolbarTextColor: request.view.toolbarTextColor,
      toolbarText: request.view.toolbarText,
      hasToolbar: request.view.hasToolbar)

    return PaymentParams(order: order, customer: customer, view: view)
  }

  private static func makeReceipt(_ message: RkReceiptMessage?) throws -> Receipt? {
    guard let message else { return nil }

    var sno: TaxSystem?
    if let raw = message.sno {
      guard let parsed = TaxSystem(rawValue: raw) else {
        throw PigeonError(
          code: "invalid_arguments",
          message: "Unknown taxation system \"\(raw)\".",
          details: nil)
      }
      sno = parsed
    }

    let items = try message.items.map { item -> ReceiptItem in
      var tax: Tax?
      if let raw = item.tax {
        guard let parsed = Tax(rawValue: raw) else {
          throw PigeonError(
            code: "invalid_arguments",
            message: "Unknown VAT rate \"\(raw)\" on receipt line \"\(item.name)\".",
            details: nil)
        }
        tax = parsed
      }
      var paymentMethod: PaymentMethod?
      if let raw = item.paymentMethod {
        guard let parsed = PaymentMethod(rawValue: raw) else {
          throw PigeonError(
            code: "invalid_arguments",
            message: "Unknown settlement method \"\(raw)\" on line \"\(item.name)\".",
            details: nil)
        }
        paymentMethod = parsed
      }
      var paymentObject: PaymentObject?
      if let raw = item.paymentObject {
        guard let parsed = PaymentObject(rawValue: raw) else {
          throw PigeonError(
            code: "invalid_arguments",
            message: "Unknown settlement subject \"\(raw)\" on line \"\(item.name)\".",
            details: nil)
        }
        paymentObject = parsed
      }

      // The SDK types `sum` as a non-optional Double, so derive the line total
      // when the caller supplied only a unit price.
      let quantity = Int(item.quantity)
      let sum = item.sum ?? ((item.cost ?? 0) * Double(quantity))

      return ReceiptItem(
        name: item.name,
        sum: sum,
        quantity: quantity,
        cost: item.cost,
        nomenclatureCode: item.nomenclatureCode,
        paymentMethod: paymentMethod,
        paymentObject: paymentObject,
        tax: tax)
    }

    return Receipt(sno: sno, items: items)
  }
}

/// Collapses the Robokassa SDK's repeat-firing closures into one answer.
private final class PaymentSession {
  private let robokassa: Robokassa
  private let invoiceId: Int64?
  /// Signed `OpStateExt` body for this invoice, or nil if there is none to ask about.
  private let checkBody: String?
  private let finish: (RkPaymentResultMessage) -> Void

  private var settled = false
  private var lastFailure: String?

  init(
    robokassa: Robokassa,
    invoiceId: Int64?,
    checkBody: String?,
    finish: @escaping (RkPaymentResultMessage) -> Void
  ) {
    self.robokassa = robokassa
    self.invoiceId = invoiceId
    self.checkBody = checkBody
    self.finish = finish
  }

  func attachHandlers() {
    robokassa.onSuccessHandler = { [weak self] opKey in
      guard let self, !self.settled else { return }
      // Set *before* the query below, on the same main-thread turn as the
      // guard: the SDK's handlers fire again while it is in flight, and this
      // flag is the only thing making those calls no-ops.
      self.settled = true
      // The SDK's success closure carries nothing but an opaque opKey, so
      // Robokassa's documented `<Result><Code>`/`<State><Code>` — which the
      // Android SDK surfaces on its own success — are read from the
      // payment-state service here. They stay nil if that query fails or times
      // out: a failed *status lookup* must not demote a payment that actually
      // succeeded.
      Task { @MainActor in
        var codes: RobokassaSdkPlugin.StateCodes?
        if let body = self.checkBody {
          do {
            codes = try await RobokassaSdkPlugin.queryStateCodes(body: body)
          } catch {
            print("Robokassa: payment succeeded but its state query failed; "
              + "resultCode/stateCode will be nil. \(error.localizedDescription)")
          }
        }
        self.finish(RkPaymentResultMessage(
          outcome: .success,
          invoiceId: self.invoiceId,
          opKey: opKey,
          resultCode: codes?.result,
          stateCode: codes?.state))
      }
    }

    robokassa.onFailureHandler = { [weak self] reason in
      // Fires repeatedly for transient "not paid yet" polls, so it only
      // records the reason; the dismissal below decides the outcome.
      guard let self, !self.settled else { return }
      self.lastFailure = reason
    }

    robokassa.onDimissHandler = { [weak self] in
      guard let self, !self.settled else { return }
      self.settled = true
      if let failure = self.lastFailure {
        self.finish(RkPaymentResultMessage(
          outcome: .error,
          invoiceId: self.invoiceId,
          errorMessage: failure))
      } else {
        self.finish(RkPaymentResultMessage(
          outcome: .canceled,
          invoiceId: self.invoiceId))
      }
    }
  }
}

File: android/src/main/kotlin/ru/robokassa/robokassa_sdk/RobokassaSdkPlugin.kt
package ru.robokassa.robokassa_sdk

import android.app.Activity
import android.content.Intent
import com.robokassa.library.models.Culture
import com.robokassa.library.models.Currency
import com.robokassa.library.models.PayActionState
import com.robokassa.library.models.PayRecurrentState
import com.robokassa.library.models.PaymentMethod
import com.robokassa.library.models.PaymentObject
import com.robokassa.library.models.Receipt
import com.robokassa.library.models.ReceiptItem
import com.robokassa.library.models.Tax
import com.robokassa.library.models.TaxSystem
import com.robokassa.library.params.CustomerParams
import com.robokassa.library.params.OrderParams
import com.robokassa.library.params.PaymentParams
import com.robokassa.library.params.ViewParams
import com.robokassa.library.pay.PaymentAction
import com.robokassa.library.pay.RobokassaPayLauncher
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.util.Date

/**
 * Bridges Flutter to Robokassa's official Android SDK.
 *
 * The checkout screen is `RobokassaActivity`, which the SDK exposes through
 * [RobokassaPayLauncher.Contract] — an `ActivityResultContract`. A Flutter
 * plugin has no `ComponentActivity` of its own to call
 * `registerForActivityResult` on, so this drives the contract manually:
 * `createIntent` to build the intent, `parseResult` to decode the outcome.
 * Reusing the SDK's own decoding beats re-reading its intent extras by hand.
 */
class RobokassaSdkPlugin :
    FlutterPlugin,
    ActivityAware,
    RobokassaHostApi,
    PluginRegistry.ActivityResultListener {

    private companion object {
        /** Arbitrary but stable; must not collide with the host app's codes. */
        const val REQUEST_CODE_PAYMENT = 41291

        const val ERROR_NO_ACTIVITY = "no_activity"
        const val ERROR_BUSY = "busy"
        const val ERROR_NATIVE = "robokassa_native_error"
    }

    private var activityBinding: ActivityPluginBinding? = null
    private var pendingPayment: ((Result<RkPaymentResultMessage>) -> Unit)? = null

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    // -------------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        RobokassaHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        RobokassaHostApi.setUp(binding.binaryMessenger, null)
        scope.cancel()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.addActivityResultListener(this)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        // A configuration change detaches and immediately re-attaches, so the
        // in-flight checkout must survive: drop the listener but keep the
        // pending callback for `onReattachedToActivityForConfigChanges`.
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
        // The result can no longer be delivered; fail the call rather than
        // leaving the Dart future pending forever.
        val callback = pendingPayment ?: return
        pendingPayment = null
        callback(
            Result.failure(
                FlutterError(
                    ERROR_NO_ACTIVITY,
                    "The host Activity went away while the Robokassa checkout was open."
                )
            )
        )
    }

    // -------------------------------------------------------------------------
    // RobokassaHostApi
    // -------------------------------------------------------------------------

    override fun isNativeSdkAvailable(): Boolean = try {
        Class.forName("com.robokassa.library.pay.RobokassaPayLauncher")
        true
    } catch (_: Throwable) {
        false
    }

    override fun startPayment(
        request: RkPaymentRequest,
        callback: (Result<RkPaymentResultMessage>) -> Unit
    ) {
        launchCheckout(request, onlyCheck = false, callback = callback)
    }

    override fun checkPaymentState(
        request: RkPaymentRequest,
        callback: (Result<RkPaymentResultMessage>) -> Unit
    ) {
        // `onlyCheck` opens RobokassaActivity with its WebView hidden: it just
        // polls the state service and returns through the same contract.
        launchCheckout(request, onlyCheck = true, callback = callback)
    }

    override fun confirmHold(request: RkPaymentRequest, callback: (Result<Boolean>) -> Unit) {
        runPaymentAction(request, callback) { action, params -> action.confirmHold(params) }
    }

    override fun cancelHold(request: RkPaymentRequest, callback: (Result<Boolean>) -> Unit) {
        runPaymentAction(request, callback) { action, params -> action.cancelHold(params) }
    }

    override fun chargeRecurring(request: RkPaymentRequest, callback: (Result<Boolean>) -> Unit) {
        runPaymentAction(request, callback) { action, params -> action.payRecurrent(params) }
    }

    // -------------------------------------------------------------------------
    // Checkout
    // -------------------------------------------------------------------------

    private fun launchCheckout(
        request: RkPaymentRequest,
        onlyCheck: Boolean,
        callback: (Result<RkPaymentResultMessage>) -> Unit
    ) {
        val activity: Activity? = activityBinding?.activity
        if (activity == null) {
            callback(
                Result.failure(
                    FlutterError(
                        ERROR_NO_ACTIVITY,
                        "Robokassa checkout needs a foreground Activity, but the plugin is " +
                            "not attached to one."
                    )
                )
            )
            return
        }
        if (pendingPayment != null) {
            callback(
                Result.failure(
                    FlutterError(
                        ERROR_BUSY,
                        "A Robokassa checkout is already in progress. Await the first call " +
                            "before starting another."
                    )
                )
            )
            return
        }

        val params = try {
            request.toPaymentParams()
        } catch (error: Throwable) {
            callback(Result.failure(FlutterError(ERROR_NATIVE, error.message ?: "$error")))
            return
        }

        pendingPayment = callback
        try {
            val intent = RobokassaPayLauncher.Contract.createIntent(
                activity,
                RobokassaPayLauncher.StartPay(
                    paymentParams = params,
                    testMode = request.credentials.isTest,
                    onlyCheck = onlyCheck
                )
            )
            activity.startActivityForResult(intent, REQUEST_CODE_PAYMENT)
        } catch (error: Throwable) {
            pendingPayment = null
            callback(
                Result.failure(
                    FlutterError(
                        ERROR_NATIVE,
                        "Could not open the Robokassa checkout screen: ${error.message ?: error}"
                    )
                )
            )
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != REQUEST_CODE_PAYMENT) return false
        val callback = pendingPayment ?: return true
        pendingPayment = null

        val message = when (
            val parsed = RobokassaPayLauncher.Contract.parseResult(resultCode, data)
        ) {
            is RobokassaPayLauncher.Success -> RkPaymentResultMessage(
                outcome = RkPaymentOutcome.SUCCESS,
                invoiceId = parsed.invoiceId?.toLong(),
                opKey = parsed.opKey,
                resultCode = parsed.resultCode?.code,
                stateCode = parsed.stateCode?.code
            )

            is RobokassaPayLauncher.Error -> RkPaymentResultMessage(
                outcome = RkPaymentOutcome.ERROR,
                resultCode = parsed.resultCode?.code,
                stateCode = parsed.stateCode?.code,
                stateDescription = parsed.desc,
                errorMessage = parsed.error?.message
                    ?: parsed.desc
                    ?: "The Robokassa checkout reported an error."
            )

            RobokassaPayLauncher.Canceled -> RkPaymentResultMessage(
                outcome = RkPaymentOutcome.CANCELED
            )
        }
        callback(Result.success(message))
        return true
    }

    // -------------------------------------------------------------------------
    // Headless operations (hold capture/release, recurring charge)
    // -------------------------------------------------------------------------

    /**
     * Runs a [PaymentAction] call and resolves [callback] exactly once.
     *
     * `PaymentAction` reports through a `StateFlow` seeded with `PayActionIdle`
     * that never completes, so the collector skips that placeholder and cancels
     * itself as soon as a real state arrives.
     */
    private fun runPaymentAction(
        request: RkPaymentRequest,
        callback: (Result<Boolean>) -> Unit,
        start: (PaymentAction, PaymentParams) -> Unit
    ) {
        val params = try {
            request.toPaymentParams()
        } catch (error: Throwable) {
            callback(Result.failure(FlutterError(ERROR_NATIVE, error.message ?: "$error")))
            return
        }

        val action = PaymentAction.init()
        var settled = false
        var job: Job? = null
        job = scope.launch {
            action.state.collect { state ->
                if (settled) return@collect
                val success = when (state) {
                    is PayActionState -> state.success
                    is PayRecurrentState -> state.success
                    // PayActionIdle — the seed value, not an answer yet.
                    else -> return@collect
                }
                settled = true
                callback(Result.success(success))
                job?.cancel()
            }
        }
        start(action, params)
    }

    // -------------------------------------------------------------------------
    // Transport records -> Robokassa SDK models
    // -------------------------------------------------------------------------

    private fun RkPaymentRequest.toPaymentParams(): PaymentParams {
        val request = this
        // `mode` is never read below — the Dart side bakes the same signal into the
        // order flags — so assert the two agree rather than let them silently drift.
        when (mode) {
            RkPaymentMode.HOLD -> require(request.order.isHold) {
                "mode=HOLD but order.isHold is false"
            }
            RkPaymentMode.RECURRENT -> require(request.order.isRecurrent) {
                "mode=RECURRENT but order.isRecurrent is false"
            }
            RkPaymentMode.SAVED_CARD -> require(!request.order.token.isNullOrEmpty()) {
                "mode=SAVED_CARD but order.token is empty"
            }
            RkPaymentMode.SIMPLE -> Unit
        }
        return PaymentParams().apply {
            order = OrderParams().apply {
                orderSum = request.order.orderSum
                // The SDK treats a non-positive id as "let Robokassa allocate".
                invoiceId = request.order.invoiceId.toInvoiceId("invoiceId")
                previousInvoiceId =
                    request.order.previousInvoiceId.toInvoiceId("previousInvoiceId")
                description = request.order.orderDescription
                incCurrLabel = request.order.incCurrLabel
                token = request.order.token
                isRecurrent = request.order.isRecurrent
                isHold = request.order.isHold
                outSumCurrency = request.order.outSumCurrency?.let(::toCurrency)
                expirationDate = request.order.expirationDateEpochMs?.let { Date(it) }
                receipt = request.order.receipt?.let(::toReceipt)
            }
            customer = CustomerParams().apply {
                culture = request.customer.culture?.let(::toCulture)
                email = request.customer.email
                ip = request.customer.ip
            }
            view = ViewParams().apply {
                hasToolbar = request.view.hasToolbar
                toolbarBgColor = request.view.toolbarBgColor
                toolbarTextColor = request.view.toolbarTextColor
                toolbarText = request.view.toolbarText
            }
            setCredentials(
                merchantLogin = request.credentials.merchantLogin,
                password1 = request.credentials.password1,
                password2 = request.credentials.password2,
                redirectUrl = request.credentials.redirectUrl
            )
        }
    }

    private fun toReceipt(message: RkReceiptMessage) = Receipt(
        sno = message.sno?.let(::toTaxSystem),
        items = message.items.map { item ->
            ReceiptItem(
                name = item.name,
                // The SDK types `sum` as a non-null Double, so derive the line
                // total when the caller supplied only a unit price.
                sum = item.sum ?: ((item.cost ?: 0.0) * item.quantity),
                quantity = item.quantity.toInt(),
                cost = item.cost,
                nomenclatureCode = item.nomenclatureCode,
                paymentMethod = item.paymentMethod?.let(::toPaymentMethod),
                paymentObject = item.paymentObject?.let(::toPaymentObject),
                tax = item.tax?.let(::toTax)
            )
        }
    )

    private fun toCulture(wire: String) = when (wire.lowercase()) {
        "en", "eng" -> Culture.EN
        else -> Culture.RU
    }

    private fun toCurrency(wire: String) = when (wire.uppercase()) {
        "USD" -> Currency.USD
        "EUR" -> Currency.EUR
        "KZT" -> Currency.KZT
        else -> Currency.RUB
    }

    private fun toTaxSystem(wire: String) = when (wire) {
        "osn" -> TaxSystem.OSN
        "usn_income" -> TaxSystem.USN_INCOME
        "usn_income_outcome" -> TaxSystem.USN_INCOME_OUTCOME
        "esn" -> TaxSystem.ESN
        "patent" -> TaxSystem.PATENT
        else -> throw IllegalArgumentException("Unknown taxation system \"$wire\"")
    }

    private fun toTax(wire: String) = when (wire) {
        "none" -> Tax.NONE
        "vat0" -> Tax.VAT_0
        "vat5" -> Tax.VAT_5
        "vat7" -> Tax.VAT_7
        "vat10" -> Tax.VAT_10
        "vat20" -> Tax.VAT_20
        "vat22" -> Tax.VAT_22
        "vat105" -> Tax.VAT_105
        "vat107" -> Tax.VAT_107
        "vat110" -> Tax.VAT_110
        "vat120" -> Tax.VAT_120
        "vat122" -> Tax.VAT_122
        else -> throw IllegalArgumentException("Unknown VAT rate \"$wire\"")
    }

    private fun toPaymentMethod(wire: String) = when (wire) {
        "full_prepayment" -> PaymentMethod.FULL_PREPAYMENT
        "prepayment" -> PaymentMethod.PREPAYMENT
        "advance" -> PaymentMethod.ADVANCE
        "full_payment" -> PaymentMethod.FULL_PAYMENT
        "partial_payment" -> PaymentMethod.PARTIAL_PAYMENT
        "credit" -> PaymentMethod.CREDIT
        "credit_payment" -> PaymentMethod.CREDIT_PAYMENT
        else -> throw IllegalArgumentException("Unknown settlement method \"$wire\"")
    }

    private fun toPaymentObject(wire: String) = when (wire) {
        "commodity" -> PaymentObject.COMMODITY
        "excise" -> PaymentObject.EXCISE
        "job" -> PaymentObject.JOB
        "service" -> PaymentObject.SERVICE
        "gambling_bet" -> PaymentObject.GAMBLING_BET
        "gambling_prize" -> PaymentObject.GAMBLING_PRIZE
        "lottery" -> PaymentObject.LOTTERY
        "lottery_prize" -> PaymentObject.LOTTERY_PRIZE
        "intellectual_activity" -> PaymentObject.INTELLECTUAL_ACTIVITY
        "payment" -> PaymentObject.PAYMENT
        "agent_commission" -> PaymentObject.AGENT_COMMISSION
        "composite" -> PaymentObject.COMPOSITE
        "resort_fee" -> PaymentObject.RESORT_FEE
        "another" -> PaymentObject.ANOTHER
        "property_right" -> PaymentObject.PROPERTY_RIGHT
        "operating_gain" -> PaymentObject.NON_OPERATING_GAIN
        "insurance_premium" -> PaymentObject.INSURANCE_PREMIUM
        "sales_tax" -> PaymentObject.SALES_TAX
        else -> throw IllegalArgumentException("Unknown settlement subject \"$wire\"")
    }
}

/**
 * Narrows a 64-bit wire invoice id to the native SDK's `Int` field.
 *
 * `null` means "let Robokassa allocate", which the SDK spells as a non-positive
 * id. An out-of-range value throws rather than silently wrapping into a
 * different — possibly valid — invoice.
 */
private fun Long?.toInvoiceId(field: String): Int {
    if (this == null) return -1
    require(this in Int.MIN_VALUE.toLong()..Int.MAX_VALUE.toLong()) {
        "$field $this does not fit in the 32-bit id Robokassa's Android SDK accepts"
    }
    return toInt()
}
